providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    public void initialize(EvaluationContext evaluationContext) throws RuntimeException {
+        try {
+            // try a dummy request
+            this.serviceBlockingStub
+                    .withWaitForReady()
+                    .withDeadlineAfter(this.deadline, TimeUnit.MILLISECONDS)
+                    .resolveBoolean(Schema.ResolveBooleanRequest.newBuilder().setFlagKey("ready?").build());
+        } catch (StatusRuntimeException e) {
+            // only return the exception if we don't meet the deadline
+            if (Status.DEADLINE_EXCEEDED.equals(e.getStatus())) {
+                throw e;
+            }
+        } finally {
+            // try in background to open the event stream
+            this.startEventStream();
+        }
+        /**
+         * Evaluate if the following makes sense
+        // we start the event stream and that will eventually configure the state to be ready.
+        // we don't need to block because the SDK run this function in a thread and doesn't join on it.
+        this.startEventStream();
+         =======
+         // Another alternative is to fail the init method on purpose and let the status recover once we are connected.
+         throw new Exception("Connecting in bg");
+         */
+    }
+
+    public void shutdown() throws Exception {
+        try {
+            if (this.channel != null) {
+                this.channel.shutdown();
+                this.channel.awaitTermination(5, TimeUnit.SECONDS);
+            }
+        } finally {
+            this.cache.clear();
+            if (this.channel != null) {
+                this.channel.shutdownNow();
+            }
+        }
+    }
+
+    public void resetRetryConnection() {
+        this.eventStreamAttempt = 1;
+        this.eventStreamRetryBackoff = this.startEventStreamRetryBackoff;
+    }
+
+    public ServiceGrpc.ServiceBlockingStub getResolver() {
+        return serviceBlockingStub.withDeadlineAfter(this.deadline, TimeUnit.MILLISECONDS);
+    }
+
+    private void startEventStream() {
+        StreamObserver<Schema.EventStreamResponse> responseObserver = new EventStreamObserver(this.cache, this.setState, this::restartEventStream);
+        this.serviceStub
+                .eventStream(Schema.EventStreamRequest.getDefaultInstance(), responseObserver);
+    }
+
+    private void restartEventStream() {
+        this.eventStreamAttempt++;
+        if (this.eventStreamAttempt > this.maxEventStreamRetries) {
+            log.error("failed to connect to event stream, exhausted retries");
+            this.setState.accept(ProviderState.ERROR);
+            return;
+        }
+        this.eventStreamRetryBackoff = 2 * this.eventStreamRetryBackoff;
+        try {
+            Thread.sleep(this.eventStreamRetryBackoff);
+        } catch (InterruptedException e) {
+            log.debug("Failed to sleep while restarting gRPC Event Stream");
+        }
+        this.startEventStream();
+    }
```

Comment:
I wanted to keep it simpler and don't move around too much critical things in a single PR. 
The circular dep will always be there to an extent since Streams don't try to reconnect by themselves. 
I planned to move the reconnect trigger from the Event Stream into the state handling of the provider. This way, the `EventStreamObserver` is responsible for only notifying that an error in the stream occurred and the provider handles the re-connection. However, I'd suggest addressing this logic in a follow-up given the complexity and here, improve the naming, WDYT @Kavindu-Dodan?

Reply:
I think a there's already a lot going on here so I'm in favor of creating another issue for this. We could even label it with `help-wanted`, etc to encourage new contributors to help with it.

Reply:
+1 for a follow-up pr

Reply:
Added to https://github.com/open-feature/java-sdk-contrib/issues/390